CC crash with recursive closure inside `inline def`

## Compiler version
3.8.3

## Minimized code

<!--
This code should be self contained, compilable (with possible failures) and as small as possible.

Ideally, we should be able to just copy this code in a file and run `scalac` (and maybe `scala`) to reproduce the issue.
-->

```Scala
package experiments.issues.suspendparams

import language.experimental.captureChecking

trait State[S, R] {
  def get(resume: S => R): R
  def put(newState: S, resume: Unit => R): R^{resume}
}

type R = (Int, Long)
def run[S1 <: State[?, R], S2 <: State[?, R]](s1: S1, s2: S2)(prog: (S1, S2) => R): R = prog(s1, s2)

inline def program(sum: State[Long, R]): R = {
  def rec(x: Int): R = {
    sum.get { s0 =>
      val s = s0 + x
      sum.put(s, _ => {
        if x > 0 then rec(x - 1) else (x, s)
      })
    }
  }
  rec(10)
}

def runProgram(other: State[?, R], sum: State[Long, R]) = {
  run(other, sum)((_, s) => program(s))
}
```

I tried the following to minimise further, but lost the error
- Removing the recursive `rec` call
- Changing `R` to a non-tuple
- Removing `R`as a parameter to State and hardcoding `(Int, Long)`
- Removing `inline` from def `program`
- Removing `^{resume}` from the return type of `State.put`
- Replacing def `run` with
  `def run[S1 <: State[?, R], S2 <: State[?, R]](s1: S1, s2: S2)(prog: S2 => R): R = prog(s2)`


## Output (click arrow to expand)
<!--
Note: to get a full crash log when using scala-cli, run compilation with the flag `--server=false`
-->
<details>

```scala
  An unhandled exception was thrown in the compiler.
  Please file a crash report here:
  https://github.com/scala/scala3/issues/new/choose
  For non-enriched exceptions, compile with -Xno-enrich-error-messages.


     while compiling: /home/pc/lab/fyp/capsicum/experiments/src/main/scala/gh-issue-snippets/SuspendParams.scala
        during phase: cc
                mode: Mode(ImplicitsEnabled)
     library version: version 3.8.3
    compiler version: version 3.8.3
            settings: -Wimplausible-patterns true -classpath /home/pc/lab/fyp/capsicum/experiments/target/scala-3.8.3/classes:/home/pc/lab/fyp/capsicum/target/scala-3.8.3/classes:/home/pc/.cache/coursier/v1/https/repo1.maven.org/maven2/org/scala-lang/scala3-library_3/3.8.3/scala3-library_3-3.8.3.jar:/home/pc/.cache/coursier/v1/https/repo1.maven.org/maven2/io/github/marcinzh/turbolift-core_3/0.124.0/turbolift-core_3-0.124.0.jar:/home/pc/.cache/coursier/v1/https/repo1.maven.org/maven2/io/github/marcinzh/beam-core_3/0.20.0/beam-core_3-0.20.0.jar:/home/pc/.cache/coursier/v1/https/repo1.maven.org/maven2/io/getkyo/kyo-prelude_3/1.0-RC1/kyo-prelude_3-1.0-RC1.jar:/home/pc/.cache/coursier/v1/https/repo1.maven.org/maven2/io/getkyo/kyo-core_3/1.0-RC1/kyo-core_3-1.0-RC1.jar:/home/pc/.cache/coursier/v1/https/repo1.maven.org/maven2/org/scala-lang/scala-library/3.8.3/scala-library-3.8.3.jar:/home/pc/.cache/coursier/v1/https/repo1.maven.org/maven2/io/getkyo/kyo-kernel_3/1.0-RC1/kyo-kernel_3-1.0-RC1.jar:/home/pc/.cache/coursier/v1/https/repo1.maven.org/maven2/io/getkyo/kyo-scheduler_3/1.0-RC1/kyo-scheduler_3-1.0-RC1.jar:/home/pc/.cache/coursier/v1/https/repo1.maven.org/maven2/org/slf4j/slf4j-api/2.0.17/slf4j-api-2.0.17.jar:/home/pc/.cache/coursier/v1/https/repo1.maven.org/maven2/dev/dirs/directories/26/directories-26.jar:/home/pc/.cache/coursier/v1/https/repo1.maven.org/maven2/io/getkyo/kyo-data_3/1.0-RC1/kyo-data_3-1.0-RC1.jar:/home/pc/.cache/coursier/v1/https/repo1.maven.org/maven2/org/jctools/jctools-core/4.0.5/jctools-core-4.0.5.jar:/home/pc/.cache/coursier/v1/https/repo1.maven.org/maven2/io/getkyo/kyo-stats-registry_3/1.0-RC1/kyo-stats-registry_3-1.0-RC1.jar:/home/pc/.cache/coursier/v1/https/repo1.maven.org/maven2/com/lihaoyi/pprint_3/0.9.0/pprint_3-0.9.0.jar:/home/pc/.cache/coursier/v1/https/repo1.maven.org/maven2/org/hdrhistogram/HdrHistogram/2.2.2/HdrHistogram-2.2.2.jar:/home/pc/.cache/coursier/v1/https/repo1.maven.org/maven2/com/lihaoyi/fansi_3/0.5.0/fansi_3-0.5.0.jar:/home/pc/.cache/coursier/v1/https/repo1.maven.org/maven2/com/lihaoyi/sourcecode_3/0.4.0/sourcecode_3-0.4.0.jar -d /home/pc/lab/fyp/capsicum/experiments/target/scala-3.8.3/classes -explain true -java-output-version 21

[error] ## Exception when compiling 12 sources to /home/pc/lab/fyp/capsicum/experiments/target/scala-3.8.3/classes
[error] java.lang.AssertionError: assertion failed
[error] scala.runtime.Scala3RunTime$.assertFailed(Scala3RunTime.scala:13)
[error] dotty.tools.dotc.core.Types$Type.toFunctionType(Types.scala:2028)
[error] dotty.tools.dotc.transform.Recheck$Rechecker.recheckClosure(Recheck.scala:423)
[error] dotty.tools.dotc.cc.CheckCaptures$CaptureChecker.recheckClosure(CheckCaptures.scala:1192)
[error] dotty.tools.dotc.cc.CheckCaptures$CaptureChecker.recheckClosureBlock(CheckCaptures.scala:1296)
[error] dotty.tools.dotc.transform.Recheck$Rechecker.recheckBlock(Recheck.scala:408)
[error] dotty.tools.dotc.transform.Recheck$Rechecker.recheckUnnamed$1(Recheck.scala:576)
[error] dotty.tools.dotc.transform.Recheck$Rechecker.recheckStart(Recheck.scala:594)
[error] dotty.tools.dotc.transform.Recheck$Rechecker.recheck(Recheck.scala:609)
[error] dotty.tools.dotc.cc.CheckCaptures$CaptureChecker.recheck(CheckCaptures.scala:1714)
[error] dotty.tools.dotc.cc.CheckCaptures$CaptureChecker.recheckArg(CheckCaptures.scala:919)
[error] dotty.tools.dotc.transform.Recheck$Rechecker.recheckArgs$1(Recheck.scala:365)
[error] dotty.tools.dotc.transform.Recheck$Rechecker.recheckApply(Recheck.scala:374)
[error] dotty.tools.dotc.cc.CheckCaptures$CaptureChecker.recheckApply(CheckCaptures.scala:908)
[error] dotty.tools.dotc.transform.Recheck$Rechecker.recheckUnnamed$1(Recheck.scala:568)
[error] dotty.tools.dotc.transform.Recheck$Rechecker.recheckStart(Recheck.scala:594)
[error] dotty.tools.dotc.transform.Recheck$Rechecker.recheck(Recheck.scala:609)
[error] dotty.tools.dotc.cc.CheckCaptures$CaptureChecker.recheck(CheckCaptures.scala:1714)
[error] dotty.tools.dotc.transform.Recheck$Rechecker.recheckBlock(Recheck.scala:403)
[error] dotty.tools.dotc.transform.Recheck$Rechecker.recheckBlock(Recheck.scala:409)
[error] dotty.tools.dotc.transform.Recheck$Rechecker.recheckUnnamed$1(Recheck.scala:576)
[error] dotty.tools.dotc.transform.Recheck$Rechecker.recheckStart(Recheck.scala:594)
[error] dotty.tools.dotc.transform.Recheck$Rechecker.recheck(Recheck.scala:609)
[error] dotty.tools.dotc.cc.CheckCaptures$CaptureChecker.recheck(CheckCaptures.scala:1714)
[error] dotty.tools.dotc.transform.Recheck$Rechecker.recheckDefDef(Recheck.scala:300)
[error] dotty.tools.dotc.cc.CheckCaptures$CaptureChecker.recheckDefDef(CheckCaptures.scala:1430)
[error] dotty.tools.dotc.transform.Recheck$Rechecker.recheckDef(Recheck.scala:536)
[error] dotty.tools.dotc.cc.CheckCaptures$CaptureChecker.completeDef(CheckCaptures.scala:1555)
[error] dotty.tools.dotc.cc.Setup$$anon$2$$anon$4.complete(Setup.scala:729)
[error] dotty.tools.dotc.core.SymDenotations$SymDenotation.completeFrom(SymDenotations.scala:176)
[error] dotty.tools.dotc.core.Denotations$Denotation.completeInfo$1(Denotations.scala:190)
[error] dotty.tools.dotc.core.Denotations$Denotation.info(Denotations.scala:192)
[error] dotty.tools.dotc.core.SymDenotations$SymDenotation.ensureCompleted(SymDenotations.scala:404)
[error] dotty.tools.dotc.transform.Recheck$Rechecker.recheckNamed$1(Recheck.scala:553)
[error] dotty.tools.dotc.transform.Recheck$Rechecker.recheckStart(Recheck.scala:593)
[error] dotty.tools.dotc.transform.Recheck$Rechecker.recheck(Recheck.scala:609)
[error] dotty.tools.dotc.cc.CheckCaptures$CaptureChecker.recheck(CheckCaptures.scala:1714)
[error] dotty.tools.dotc.cc.CheckCaptures$CaptureChecker.recheckStat(CheckCaptures.scala:1694)
[error] dotty.tools.dotc.transform.Recheck$Rechecker.traverse$1(Recheck.scala:522)
[error] dotty.tools.dotc.transform.Recheck$Rechecker.recheckStats(Recheck.scala:525)
[error] dotty.tools.dotc.transform.Recheck$Rechecker.recheckClassDef(Recheck.scala:310)
[error] dotty.tools.dotc.cc.CheckCaptures$CaptureChecker.recheckClassDef(CheckCaptures.scala:1622)
[error] dotty.tools.dotc.transform.Recheck$Rechecker.recheckNamed$1(Recheck.scala:561)
[error] dotty.tools.dotc.transform.Recheck$Rechecker.recheckStart(Recheck.scala:593)
[error] dotty.tools.dotc.transform.Recheck$Rechecker.recheck(Recheck.scala:609)
[error] dotty.tools.dotc.cc.CheckCaptures$CaptureChecker.recheck(CheckCaptures.scala:1714)
[error] dotty.tools.dotc.cc.CheckCaptures$CaptureChecker.recheckStat(CheckCaptures.scala:1694)
[error] dotty.tools.dotc.transform.Recheck$Rechecker.traverse$1(Recheck.scala:522)
[error] dotty.tools.dotc.transform.Recheck$Rechecker.recheckStats(Recheck.scala:525)
[error] dotty.tools.dotc.transform.Recheck$Rechecker.recheckPackageDef(Recheck.scala:511)
[error] dotty.tools.dotc.transform.Recheck$Rechecker.recheckUnnamed$1(Recheck.scala:588)
[error] dotty.tools.dotc.transform.Recheck$Rechecker.recheckStart(Recheck.scala:594)
[error] dotty.tools.dotc.transform.Recheck$Rechecker.recheck(Recheck.scala:609)
[error] dotty.tools.dotc.cc.CheckCaptures$CaptureChecker.recheck(CheckCaptures.scala:1714)
[error] dotty.tools.dotc.transform.Recheck$Rechecker.checkUnit(Recheck.scala:667)
[error] dotty.tools.dotc.cc.CheckCaptures$CaptureChecker.iterate$1(CheckCaptures.scala:2231)
[error] dotty.tools.dotc.cc.CheckCaptures$CaptureChecker.checkUnit$$anonfun$3(CheckCaptures.scala:2242)
[error] scala.runtime.function.JProcedure1.apply(JProcedure1.java:15)
[error] scala.runtime.function.JProcedure1.apply(JProcedure1.java:10)
[error] dotty.tools.dotc.cc.CaptureSet$.withCaptureSetsExplained(CaptureSet.scala:1847)
[error] dotty.tools.dotc.cc.CheckCaptures$CaptureChecker.checkUnit(CheckCaptures.scala:2248)
[error] dotty.tools.dotc.transform.Recheck.run(Recheck.scala:154)
[error] dotty.tools.dotc.cc.CheckCaptures.run(CheckCaptures.scala:276)
[error] dotty.tools.dotc.core.Phases$Phase.runOn$$anonfun$1(Phases.scala:406)
[error] scala.runtime.function.JProcedure1.apply(JProcedure1.java:15)
[error] scala.runtime.function.JProcedure1.apply(JProcedure1.java:10)
[error] scala.collection.immutable.List.foreach(List.scala:327)
[error] dotty.tools.dotc.core.Phases$Phase.runOn(Phases.scala:399)
[error] dotty.tools.dotc.transform.Recheck.runOn(Recheck.scala:158)
[error] dotty.tools.dotc.cc.CheckCaptures.runOn(CheckCaptures.scala:272)
[error] dotty.tools.dotc.Run.runPhases$1$$anonfun$1(Run.scala:380)
[error] scala.runtime.function.JProcedure1.apply(JProcedure1.java:15)
[error] scala.runtime.function.JProcedure1.apply(JProcedure1.java:10)
[error] scala.collection.ArrayOps$.foreach$extension(ArrayOps.scala:1324)
[error] dotty.tools.dotc.Run.runPhases$1(Run.scala:373)
[error] dotty.tools.dotc.Run.compileUnits$$anonfun$1$$anonfun$2(Run.scala:420)
[error] dotty.tools.dotc.Run.compileUnits$$anonfun$1$$anonfun$adapted$1(Run.scala:420)
[error] scala.Function0.apply$mcV$sp(Function0.scala:42)
[error] dotty.tools.dotc.Run.showProgress(Run.scala:482)
[error] dotty.tools.dotc.Run.compileUnits$$anonfun$1(Run.scala:420)
[error] dotty.tools.dotc.Run.compileUnits$$anonfun$adapted$1(Run.scala:432)
[error] dotty.tools.dotc.util.Stats$.maybeMonitored(Stats.scala:69)
[error] dotty.tools.dotc.Run.compileUnits(Run.scala:432)
[error] dotty.tools.dotc.Run.compileSources(Run.scala:319)
[error] dotty.tools.dotc.Run.compile(Run.scala:304)
[error] dotty.tools.dotc.Driver.doCompile(Driver.scala:37)
[error] dotty.tools.xsbt.CompilerBridgeDriver.run(CompilerBridgeDriver.java:141)
[error] dotty.tools.xsbt.CompilerBridge.run(CompilerBridge.java:22)
[error] sbt.internal.inc.AnalyzingCompiler.compile(AnalyzingCompiler.scala:91)
[error] sbt.internal.inc.MixedAnalyzingCompiler.$anonfun$compile$7(MixedAnalyzingCompiler.scala:196)
[error] scala.runtime.java8.JFunction0$mcV$sp.apply(JFunction0$mcV$sp.java:23)
[error] sbt.internal.inc.MixedAnalyzingCompiler.timed(MixedAnalyzingCompiler.scala:252)
[error] sbt.internal.inc.MixedAnalyzingCompiler.$anonfun$compile$4(MixedAnalyzingCompiler.scala:186)
[error] sbt.internal.inc.MixedAnalyzingCompiler.$anonfun$compile$4$adapted(MixedAnalyzingCompiler.scala:166)
[error] sbt.internal.inc.JarUtils$.withPreviousJar(JarUtils.scala:241)
[error] sbt.internal.inc.MixedAnalyzingCompiler.compileScala$1(MixedAnalyzingCompiler.scala:166)
[error] sbt.internal.inc.MixedAnalyzingCompiler.compile(MixedAnalyzingCompiler.scala:214)
[error] sbt.internal.inc.IncrementalCompilerImpl.$anonfun$compileInternal$1(IncrementalCompilerImpl.scala:542)
[error] sbt.internal.inc.IncrementalCompilerImpl.$anonfun$compileInternal$1$adapted(IncrementalCompilerImpl.scala:542)
[error] sbt.internal.inc.Incremental$.$anonfun$apply$3(Incremental.scala:182)
[error] sbt.internal.inc.Incremental$.$anonfun$apply$3$adapted(Incremental.scala:180)
[error] sbt.internal.inc.Incremental$$anon$2.run(Incremental.scala:458)
[error] sbt.internal.inc.IncrementalCommon$CycleState.next(IncrementalCommon.scala:116)
[error] sbt.internal.inc.IncrementalCommon$$anon$1.next(IncrementalCommon.scala:56)
[error] sbt.internal.inc.IncrementalCommon$$anon$1.next(IncrementalCommon.scala:52)
[error] sbt.internal.inc.IncrementalCommon.cycle(IncrementalCommon.scala:264)
[error] sbt.internal.inc.Incremental$.$anonfun$incrementalCompile$8(Incremental.scala:413)
[error] sbt.internal.inc.Incremental$.withClassfileManager(Incremental.scala:500)
[error] sbt.internal.inc.Incremental$.incrementalCompile(Incremental.scala:400)
[error] sbt.internal.inc.Incremental$.apply(Incremental.scala:208)
[error] sbt.internal.inc.IncrementalCompilerImpl.compileInternal(IncrementalCompilerImpl.scala:542)
[error] sbt.internal.inc.IncrementalCompilerImpl.$anonfun$compileIncrementally$1(IncrementalCompilerImpl.scala:496)
[error] sbt.internal.inc.IncrementalCompilerImpl.handleCompilationError(IncrementalCompilerImpl.scala:332)
[error] sbt.internal.inc.IncrementalCompilerImpl.compileIncrementally(IncrementalCompilerImpl.scala:433)
[error] sbt.internal.inc.IncrementalCompilerImpl.compile(IncrementalCompilerImpl.scala:137)
[error] sbt.Defaults$.compileIncrementalTaskImpl(Defaults.scala:2340)
[error] sbt.Defaults$.$anonfun$compileIncrementalTask$2(Defaults.scala:2290)
[error] sbt.internal.server.BspCompileTask$.$anonfun$compute$1(BspCompileTask.scala:41)
[error] sbt.internal.io.Retry$.sbt$internal$io$Retry$$impl(Retry.scala:114)
[error] sbt.internal.io.Retry$.io(Retry.scala:102)
[error] sbt.internal.io.Retry$.io(Retry.scala:78)
[error] sbt.internal.io.Retry$.io(Retry.scala:67)
[error] sbt.internal.server.BspCompileTask$.compute(BspCompileTask.scala:41)
[error] sbt.Defaults$.$anonfun$compileIncrementalTask$1(Defaults.scala:2288)
[error] scala.Function1.$anonfun$compose$1(Function1.scala:49)
[error] sbt.internal.util.$tilde$greater.$anonfun$$u2219$1(TypeFunctions.scala:63)
[error] sbt.std.Transform$$anon$4.work(Transform.scala:69)
[error] sbt.Execute.$anonfun$submit$2(Execute.scala:283)
[error] sbt.internal.util.ErrorHandling$.wideConvert(ErrorHandling.scala:24)
[error] sbt.Execute.work(Execute.scala:292)
[error] sbt.Execute.$anonfun$submit$1(Execute.scala:283)
[error] sbt.ConcurrentRestrictions$$anon$4.$anonfun$submitValid$1(ConcurrentRestrictions.scala:265)
[error] sbt.CompletionService$$anon$2.call(CompletionService.scala:65)
[error] java.base/java.util.concurrent.FutureTask.run(FutureTask.java:317)
[error] java.base/java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:572)
[error] java.base/java.util.concurrent.FutureTask.run(FutureTask.java:317)
[error] java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1144)
[error] java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:642)
[error] java.base/java.lang.Thread.run(Thread.java:1583)
[error]            
[error] stack trace is suppressed; run last Compile / compileIncremental for the full output
[error] (Compile / compileIncremental) java.lang.AssertionError: assertion failed
```
</details>

--- a/tests/pos-custom-args/captures/i26186.scala
+++ b/tests/pos-custom-args/captures/i26186.scala
@@ -1,0 +1,27 @@
+package experiments.issues.suspendparams
+
+import language.experimental.captureChecking
+
+trait State[S, R] {
+  def get(resume: S => R): R
+  def put(newState: S, resume: Unit => R): R^{resume}
+}
+
+type R = (Int, Long)
+def run[S1 <: State[?, R], S2 <: State[?, R]](s1: S1, s2: S2)(prog: (S1, S2) => R): R = prog(s1, s2)
+
+inline def program(sum: State[Long, R]): R = {
+  def rec(x: Int): R = {
+    sum.get { s0 =>
+      val s = s0 + x
+      sum.put(s, _ => {
+        if x > 0 then rec(x - 1) else (x, s)
+      })
+    }
+  }
+  rec(10)
+}
+
+def runProgram(other: State[?, R], sum: State[Long, R]) = {
+  run(other, sum)((_, s) => program(s))
+}

